src/js/bs3/module/HintPopover.js

```diff
@@ -0,0 +1,273 @@
+        });
+      }
+    }).render().appendTo('body');
+
+    var $popoverContent = $popover.find('.popover-content');
+
+
+    this.timer = null;
+
+
+    this.events = {
+      'summernote.keyup': function (e, nativeEvent) {
+        self.update(nativeEvent);
+      },
+      'summernote.keydown' : function (e, nativeEvent) {
+        self.updateKeydown(nativeEvent);
+      }
+    };
+
+    this.initialize = function () {
+      if (!hint) {
+        return;
+      }
+      dom.attachEvents($note, this.events);
+
+      $popoverContent.on('click', '.hint-item', function () {
+        $popoverContent.find('.active').removeClass('active');
+        $(this).addClass('active');
+        self.replace();
+        self.hide();
+      });
+    };
+
+    this.destroy = function () {
+      if (!hint) {
+        return;
+      }
+      dom.detachEvents($note, this.events);
+      $popoverContent.off('click');
+    };
+
+    this.activate = function (item) {
+      var $activeItem = $(item);
+      $popoverContent.find('.active').removeClass('active');
+      $activeItem.addClass('active');
+
+      this.scrollTo($activeItem);
+    };
+
+    this.scrollTo = function ($node) {
+      var $parent = $node.parent().parent();
+      $parent[0].scrollTop = $node[0].offsetTop - ($parent.innerHeight() / 2);
+    };
+
+    this.moveDown = function () {
+      var $old = $popoverContent.find('.hint-item.active');
+      var $next = $old.next();
+
+      if ($next.length) {
+        this.activate($next);
+      } else {
+        var $parentNext = $old.parent().next();
+
+        if (!$parentNext.length) {
+          $parentNext = $popoverContent.find('.hint-group').first();
+        }
+
+        this.activate($($parentNext).find('.hint-item').first());
+      }
+    };
+
+    this.moveUp = function () {
+      var $old = $popoverContent.find('.hint-item.active');
+      var $prev = $old.prev();
+
+      if ($prev.length) {
+        this.activate($prev);
+      } else {
+        var $parentPrev = $old.parent().prev();
+
+        if (!$parentPrev.length) {
+          $parentPrev = $popoverContent.find('.hint-group').last();
+        }
+
+        this.activate($($parentPrev).find('.hint-item').last());
+      }
+    };
+
+    this.replace = function () {
+      var wordRange = $popover.data('wordRange');
+      var $activeItem = $popoverContent.find('.active');
+      var content = this.content($activeItem.data('index'), $activeItem.data('item'));
+
+      if (typeof content === 'string') {
+        content = document.createTextNode(content);
+      }
+
+      $popover.removeData('wordRange');
+
+      wordRange.insertNode(content);
+      range.createFromNode(content).collapse().select();
+    };
+
+    this.content = function (index, obj) {
+      if (hint[index] && hint[index].content) {
+        return hint[index].content(obj);
+      } else {
+        return obj;
+      }
+    };
+
+    this.searchKeyword = function (hint, keyword, callback) {
```

Comment:
`hint` is already defined at function scope. Do you have any reason for passing hint?